modules/aws-http4s/src/smithy4s/aws/internals/AwsQueryCodecs.scala

```diff
@@ -0,0 +1,226 @@
+/*
+ *  Copyright 2023 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+package aws
+package internals
+
+import cats.effect.Concurrent
+import cats.syntax.all._
+import fs2.compression.Compression
+import smithy4s.Endpoint
+import smithy4s.schema.CachedSchemaCompiler
+import org.http4s.EntityDecoder
+import smithy4s.xml.XmlDocument
+import org.http4s.MediaRange
+import fs2.data.xml._
+import fs2.data.xml.dom._
+import smithy4s.http._
+import smithy4s.http4s.kernel._
+import cats.data.EitherT
+import smithy4s.kinds.PolyFunction
+import smithy4s.codecs.PayloadPath
+import org.http4s.EntityEncoder
+import smithy4s.capability.Covariant
+import _root_.aws.protocols.AwsQueryError
+
+import smithy4s.http.Metadata
+
+private[aws] object AwsQueryCodecs {
+
+  def make[F[_]: Concurrent: Compression](
+      version: String
+  ): UnaryClientCodecs.Make[F] =
+    new UnaryClientCodecs.Make[F] {
+      def apply[I, E, O, SI, SO](
+          endpoint: Endpoint.Base[I, E, O, SI, SO]
+      ): UnaryClientCodecs[F, I, E, O] = {
+
+        val stringAndBlobsEntityDecoders =
+          smithy4s.http.StringAndBlobCodecs.ReaderCompiler
+            .mapK(
+              Covariant.liftPolyFunction[Option](
+                EntityDecoders.fromHttpMediaReaderK[F]
+              )
+            )
+
+        val urlFormEntityEncoders: CachedSchemaCompiler[EntityEncoder[F, *]] =
+          UrlForm.Encoder
+            .mapK(
+              new PolyFunction[UrlForm.Encoder, EntityEncoder[F, *]] {
+                def apply[A](fa: UrlForm.Encoder[A]): EntityEncoder[F, A] =
+                  urlFormEntityEncoder[F].contramap((a: A) =>
+                    UrlForm(
+                      UrlForm.FormData.MultipleValues(
+                        UrlForm.FormData
+                          .PathedValue(
+                            PayloadPath(PayloadPath.Segment("Action")),
+                            endpoint.id.name
+                          )
+                          .toPathedValues ++
+                          UrlForm.FormData
+                            .PathedValue(
+                              PayloadPath(PayloadPath.Segment("Version")),
+                              version
+                            )
+                            .toPathedValues ++
+                          fa.encode(a).values.values
```

Comment:
aesthetically speaking, I'm not a fan of [hadouken code](https://i.pinimg.com/originals/98/bd/38/98bd38321fad0e8b14ebf3a5a9ec70c9.jpg) 

I'd rather the `Action`/`Version` related was extracted in a `addOperationHeader` method

Reply:
Yeah this mess isn't meant to be final, it's just another step along the way.